internal/satellite/module/buffer/buffer.go

```diff
@@ -0,0 +1,94 @@
+// Licensed to Apache Software Foundation (ASF) under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Apache Software Foundation (ASF) licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package buffer
+
+import (
+	"sync"
+
+	"github.com/apache/skywalking-satellite/internal/pkg/event"
+	"github.com/apache/skywalking-satellite/internal/pkg/log"
+)
+
+// BatchBuffer is a buffer to cache the input data in Sender.
+type BatchBuffer struct {
+	sync.Mutex                             // local
+	buf        []*event.OutputEventContext // cache
+	first      int64                       // the first OutputEventContext offset
+	last       int64                       // the last OutputEventContext offset
+	size       int                         // usage size
+	cap        int                         // the max capacity
+}
+
+// NewBuffer crate a new BatchBuffer according to the capacity param.
+func NewBatchBuffer(capacity int) *BatchBuffer {
+	return &BatchBuffer{
+		buf:   make([]*event.OutputEventContext, capacity),
+		first: 0,
+		last:  0,
+		size:  0,
+		cap:   capacity,
+	}
+}
+
+// Buf returns the cached data in BatchBuffer.
+func (b *BatchBuffer) Buf() []*event.OutputEventContext {
+	b.Lock()
```

Comment:
so many locks. it may leads to the system slowing down ？

Reply:
Yes, I want to say this. Is this really necessary for using lock even in the register? Who will use concurrency-registration?

Reply:
Any update?

Reply:
I am polishing it now